File: packages/hermes-mybot-farm/farm_api.py
"""Live mybot.farm API helpers (stdlib only)."""

from __future__ import annotations

import json
import os
from typing import Any
from urllib.error import HTTPError, URLError
from urllib.parse import quote, urljoin
from urllib.request import Request, urlopen

DEFAULT_BASE = "https://mybot.farm"
USER_AGENT = "hermes-mybot-farm/0.2.0"
TIMEOUT_S = 60

# Exact category *labels* from web/src/lib/site.ts `categories[].label`.
CATEGORY_LABELS = frozenset(
    {
        "Lifestyle",
        "Productivity",
        "Coding",
        "Writing",
        "Marketing",
        "Sales",
        "Research",
        "Personal finance",
        "Creative",
        "Music",
        "Education",
        "Ops / admin",
        "Experimental",
    }
)
LISTING_KINDS = frozenset({"agent", "team"})
AGENT_PACK_FORMAT = "mybot.farm/agent-pack"
TEAM_PACK_FORMAT = "mybot.farm/team-pack"
MIN_TEAM_MEMBERS = 2
MIN_PAID_PRICE_CENTS = 200
MAX_PRICE_CENTS = 999_900
MAX_PACK_CHARS = 500_000
PRICE_HINT = "Choose Free, or a price between $2.00 and $9,999.00."


class FarmError(Exception):
    """HTTP or pack-shape error from mybot.farm."""

    def __init__(self, message: str, status: int | None = None) -> None:
        super().__init__(message)
        self.status = status


def resolve_base_url(plugin_config: dict[str, Any] | None = None) -> str:
    env = (os.environ.get("MYBOT_FARM_URL") or "").strip()
    cfg = plugin_config or {}
    cfg_url = str(cfg.get("baseUrl") or cfg.get("base_url") or "").strip()
    base = env or cfg_url or DEFAULT_BASE
    return base.rstrip("/")


def resolve_api_key(
    plugin_config: dict[str, Any] | None = None,
    override: str | None = None,
) -> str:
    """Seller key: per-call override, else env MYBOT_FARM_API_KEY, else config apiKey.

    Never log the returned value.
    """
    if isinstance(override, str) and override.strip():
        return override.strip()
    env = (os.environ.get("MYBOT_FARM_API_KEY") or "").strip()
    if env:
        return env
    cfg = plugin_config or {}
    return str(cfg.get("apiKey") or cfg.get("api_key") or "").strip()


def _short_error_body(raw: str) -> str:
    text = (raw or "").strip()
    if not text:
        return ""
    try:
        data = json.loads(text)
    except json.JSONDecodeError:
        return text[:200]
    if not isinstance(data, dict):
        return text[:200]
    parts: list[str] = []
    for key in ("error", "message"):
        value = data.get(key)
        if isinstance(value, str) and value.strip():
            parts.append(value.strip())
    return ": ".join(parts)[:200] if parts else text[:200]


def _request(
    url: str,
    *,
    accept: str = "application/json",
    data: bytes | None = None,
    headers: dict[str, str] | None = None,
    method: str | None = None,
) -> bytes:
    req_headers = {
        "Accept": accept,
        "User-Agent": USER_AGENT,
    }
    if headers:
        req_headers.update(headers)
    req = Request(url, data=data, headers=req_headers, method=method)
    try:
        with urlopen(req, timeout=TIMEOUT_S) as resp:
            return resp.read()
    except HTTPError as exc:
        body = ""
        try:
            body = exc.read().decode("utf-8", errors="replace")[:240]
        except Exception:
            body = ""
        short = _short_error_body(body)
        raise FarmError(
            f"mybot.farm {exc.code} for {url}{(': ' + short) if short else ''}",
            status=exc.code,
        ) from exc
    except URLError as exc:
        raise FarmError(f"mybot.farm unreachable for {url}: {exc.reason}") from exc


def farm_fetch_json(base_url: str, path: str) -> Any:
    url = f"{base_url}{path if path.startswith('/') else '/' + path}"
    raw = _request(url)
    try:
        return json.loads(raw.decode("utf-8"))
    except json.JSONDecodeError as exc:
        raise FarmError(f"non-JSON from {url}: {exc}") from exc


def farm_fetch_bytes(url: str) -> bytes:
    return _request(url, accept="*/*")


def search_stalls(base_url: str, query: str, limit: int | None = None) -> dict[str, Any]:
    q = quote(query, safe="")
    data = farm_fetch_json(base_url, f"/api/stalls?q={q}")
    stalls = data.get("stalls") if isinstance(data.get("stalls"), list) else []
    if isinstance(limit, int) and limit > 0:
        stalls = stalls[:limit]
    count = data.get("count") if isinstance(data.get("count"), int) else len(stalls)
    return {"stalls": stalls, "count": count, "query": data.get("query", query)}


def get_stall(base_url: str, slug: str) -> dict[str, Any]:
    clean = slug.strip().strip("/")
    if not clean:
        raise FarmError("slug required")
    data = farm_fetch_json(base_url, f"/api/stalls/{quote(clean, safe='')}")
    if not isinstance(data, dict):
        raise FarmError(f"empty stall for {clean}")
    data.setdefault("slug", clean)
    return data


def get_pack(base_url: str, slug: str) -> dict[str, Any]:
    clean = slug.strip().strip("/")
    if not clean:
        raise FarmError("slug required")
    pack = farm_fetch_json(base_url, f"/api/packs/{quote(clean, safe='')}")
    if not isinstance(pack, dict):
        raise FarmError(f"empty pack for {clean}")
    pack.setdefault("slug", clean)
    return pack


def absolute_url(base_url: str, href: str) -> str:
    if href.startswith("http://") or href.startswith("https://"):
        return href
    return urljoin(base_url + "/", href.lstrip("/"))


def is_hermes_archive(path: str | None) -> bool:
    if not path:
        return False
    lower = path.split("?", 1)[0].lower()
    return lower.endswith(".hermes.tar.gz") or lower.endswith(".tar.gz")


def _runtime_list(*sources: Any) -> list[str]:
    values: list[str] = []
    for source in sources:
        if isinstance(source, list):
            values.extend(str(item) for item in source)
    return values


def resolve_agent_archive_href(stall: dict[str, Any], pack: dict[str, Any]) -> str:
    """Prefer stall.hermesHref, then a .tar.gz downloadHref, then /packs/{kind}/{slug}.hermes.tar.gz."""
    for key in ("hermesHref", "hermesUrl"):
        value = stall.get(key)
        if is_hermes_archive(str(value or "")):
            return str(value)
    download_href = str(stall.get("downloadHref") or stall.get("packUrl") or "")
    if is_hermes_archive(download_href):
        return download_href
    nested = stall.get("pack") if isinstance(stall.get("pack"), dict) else {}
    runtimes = _runtime_list(pack.get("runtime"), nested.get("runtime"))
    if "hermes" not in {item.lower() for item in runtimes}:
        return ""
    slug = str(pack.get("slug") or stall.get("slug") or "")
    if not slug:
        return ""
    kind = stall.get("kind") or (
        "team" if str(pack.get("format") or "").endswith("team-pack") else "agent"
    )
    folder = "teams" if kind == "team" else "agents"
    return f"/packs/{folder}/{slug}.hermes.tar.gz"


def stall_summary(stall: dict[str, Any]) -> dict[str, Any]:
    slug = stall.get("slug") or ""
    nested_pack = stall.get("pack") if isinstance(stall.get("pack"), dict) else {}
    pack_version = stall.get("packVersion")
    if pack_version is None:
        pack_version = nested_pack.get("packVersion")
    return {
        "slug": slug,
        "stallId": stall.get("stallId") or stall.get("listingId") or "",
        "packVersion": pack_version,
        "kind": stall.get("kind") or "",
        "name": stall.get("name") or slug,
        "title": stall.get("title") or "",
        "pageUrl": stall.get("pageUrl") or f"{DEFAULT_BASE}/agents/{slug}",
        "packUrl": stall.get("packUrl") or "",
        "downloadHref": stall.get("downloadHref") or "",
        "hermesHref": stall.get("hermesHref") or "",
        "hermesUrl": stall.get("hermesUrl") or "",
        "description": stall.get("description") or "",
        "category": stall.get("category") or "",
        "members": stall.get("members") or [],
        "runtime": (stall.get("pack") or {}).get("runtime")
        if isinstance(stall.get("pack"), dict)
        else [],
    }


def pack_summary(pack: dict[str, Any]) -> dict[str, Any]:
    skills = pack.get("skills") if isinstance(pack.get("skills"), list) else []
    members = pack.get("members") if isinstance(pack.get("members"), list) else []
    manifest = pack.get("manifest") if isinstance(pack.get("manifest"), dict) else {}
    shared = pack.get("shared") if isinstance(pack.get("shared"), dict) else {}
    getting = shared.get("gettingStarted")
    if not isinstance(getting, str):
        getting = ""
    return {
        "slug": pack.get("slug"),
        "format": pack.get("format"),
        "version": pack.get("version"),
        "packVersion": pack.get("packVersion"),
        "runtime": pack.get("runtime") or [],
        "profile": pack.get("profile") or {},
        "skillNames": [s.get("name") for s in skills if isinstance(s, dict) and s.get("name")],
        "skillCount": len(skills),
        "memberCount": len(members),
        "members": [
            {
                "role": m.get("role"),
                "summary": m.get("summary"),
                "pack": m.get("pack"),
                "slug": m.get("slug"),
            }
            for m in members
            if isinstance(m, dict)
        ],
        "gettingStarted": getting,
        "attribution": manifest.get("attribution") or "",
        "sourceNote": manifest.get("sourceNote") or "",
        "license": manifest.get("license") or "",
        "homepage": manifest.get("homepage") or "",
        "scrubbed": manifest.get("scrubbed"),
    }


def parse_price_cents(value: Any) -> int | None:
    """Match web/src/lib/listings.ts parsePriceCents (plus string ints for CLI)."""
    if isinstance(value, bool):
        return None
    if isinstance(value, str):
        text = value.strip()
        if not text:
            return None
        try:
            value = float(text) if "." in text else int(text)
        except ValueError:
            return None
    if not isinstance(value, (int, float)):
        return None
    if isinstance(value, float) and (value != value or value in {float("inf"), float("-inf")}):
        return None
    cents = int(round(value))
    if cents == 0:
        return 0
    if cents < MIN_PAID_PRICE_CENTS or cents > MAX_PRICE_CENTS:
        return None
    return cents


def parse_listing_kind(value: Any) -> str | None:
    if isinstance(value, str) and value.strip() in LISTING_KINDS:
        return value.strip()
    return None


def parse_pack_object(value: Any) -> dict[str, Any]:
    if isinstance(value, str):
        try:
            value = json.loads(value)
        except json.JSONDecodeError as exc:
            raise FarmError("Pack JSON must be an object.") from exc
    if not isinstance(value, dict) or isinstance(value, list):
        raise FarmError("Pack JSON must be an object.")
    encoded = json.dumps(value, ensure_ascii=False, separators=(",", ":"))
    if len(encoded) > MAX_PACK_CHARS:
        raise FarmError("Pack JSON is too large (max 500 KB).")
    return value


def _member_pack_error(index: int, pack: Any) -> str | None:
    if isinstance(pack, str):
        return (
            None
            if pack.strip()
            else (
                f"members[{index}].pack must be a catalog path, slug, tarball URL, "
                "or nested agent-pack object"
            )
        )
    if not isinstance(pack, dict) or isinstance(pack, list):
        return (
            f"members[{index}].pack must be a catalog path, slug, tarball URL, "
            "or nested agent-pack object"
        )
    if pack.get("format") == TEAM_PACK_FORMAT:
        return f"members[{index}].pack nested object cannot be a team-pack"
    return None


def validate_listing_pack(kind: str, pack: dict[str, Any]) -> None:
    """Match web/src/lib/gaf-pack.ts validateListingPack for farm_post dry-run."""
    fmt = pack.get("format").strip() if isinstance(pack.get("format"), str) else ""
    if kind == "team":
        if fmt != TEAM_PACK_FORMAT:
            raise FarmError(f'kind "team" requires pack.format "{TEAM_PACK_FORMAT}"')
        members = pack.get("members")
        if not isinstance(members, list) or len(members) < MIN_TEAM_MEMBERS:
            raise FarmError(
                f'kind "team" requires members[] with at least {MIN_TEAM_MEMBERS} agents'
            )
        for i, member in enumerate(members):
            if not isinstance(member, dict) or isinstance(member, list):
                raise FarmError(
                    f"members[{i}] must be an object with role, summary, and pack"
                )
            role = member.get("role").strip() if isinstance(member.get("role"), str) else ""
            summary = (
                member.get("summary").strip()
                if isinstance(member.get("summary"), str)
                else ""
            )
            if not role:
                raise FarmError(f"members[{i}].role is required")
            if not summary:
                raise FarmError(f"members[{i}].summary is required")
            pack_error = _member_pack_error(i, member.get("pack"))
            if pack_error:
                raise FarmError(pack_error)
        shared = pack.get("shared")
        if shared is not None:
            if not isinstance(shared, dict) or isinstance(shared, list):
                raise FarmError("shared must be an object.")
            getting = shared.get("gettingStarted")
            if getting is not None and not isinstance(getting, str):
                raise FarmError(
                    "shared.gettingStarted must be a string (Hermes install steps)."
                )
        return
    if fmt == TEAM_PACK_FORMAT:
        raise FarmError(f'kind "agent" cannot use pack.format "{TEAM_PACK_FORMAT}"')
    members = pack.get("members")
    if isinstance(members, list) and members:
        raise FarmError('kind "agent" listings cannot include members[] — use kind "team"')


def build_listing_payload(
    *,
    kind: Any,
    name: Any,
    title: Any,
    description: Any,
    category: Any,
    price_cents: Any,
    pack: Any,
    slug: Any = None,
    pack_version: Any = None,
) -> dict[str, Any]:
    parsed_kind = parse_listing_kind(kind)
    if not parsed_kind:
        raise FarmError('kind must be "agent" or "team"')

    parsed_name = name.strip() if isinstance(name, str) else ""
    parsed_title = title.strip() if isinstance(title, str) else ""
    parsed_description = description.strip() if isinstance(description, str) else ""
    if not parsed_name or not parsed_title or not parsed_description:
        raise FarmError("name, title, and description are required")

    parsed_category = category.strip() if isinstance(category, str) else ""
    if parsed_category not in CATEGORY_LABELS:
        labels = ", ".join(sorted(CATEGORY_LABELS))
        raise FarmError(f"category must be an exact farm label ({labels})")

    parsed_price = parse_price_cents(price_cents)
    if parsed_price is None:
        raise FarmError(PRICE_HINT)

    parsed_pack = parse_pack_object(pack)
    validate_listing_pack(parsed_kind, parsed_pack)
    payload: dict[str, Any] = {
        "kind": parsed_kind,
        "name": parsed_name,
        "title": parsed_title,
        "description": parsed_description,
        "category": parsed_category,
        "priceCents": parsed_price,
        "pack": parsed_pack,
    }
    if isinstance(slug, str) and slug.strip():
        payload["slug"] = slug.strip().lower()
    if pack_version is not None and pack_version != "":
        payload["packVersion"] = pack_version
    return payload


def listing_payload_summary(payload: dict[str, Any]) -> dict[str, Any]:
    pack = payload.get("pack") if isinstance(payload.get("pack"), dict) else {}
    skills = pack.get("skills") if isinstance(pack.get("skills"), list) else []
    members = pack.get("members") if isinstance(pack.get("members"), list) else []
    encoded = json.dumps(pack, ensure_ascii=False, separators=(",", ":"))
    roles = [
        str(m.get("role")).strip()
        for m in members
        if isinstance(m, dict) and isinstance(m.get("role"), str) and str(m.get("role")).strip()
    ]
    return {
        "kind": payload.get("kind"),
        "name": payload.get("name"),
        "title": payload.get("title"),
        "category": payload.get("category"),
        "priceCents": payload.get("priceCents"),
        "slug": payload.get("slug"),
        "packVersion": payload.get("packVersion"),
        "pack": {
            "format": pack.get("format"),
            "version": pack.get("version"),
            "packVersion": pack.get("packVersion"),
            "runtime": pack.get("runtime") or [],
            "skillCount": len(skills),
            "memberCount": len(members),
            "memberRoles": roles,
            "encodedChars": len(encoded),
        },
    }


def listing_page_url(base_url: str, page_path: str) -> str:
    path = page_path if page_path.startswith("/") else f"/{page_path}"
    return f"{base_url.rstrip('/')}{path}"


def create_listing(
    base_url: str,
    payload: dict[str, Any],
    api_key: str,
) -> dict[str, Any]:
    key = (api_key or "").strip()
    if not key:
        raise FarmError(
            "seller API key required (env MYBOT_FARM_API_KEY or plugin config apiKey)"
        )
    url = f"{base_url.rstrip('/')}/api/listings"
    body = json.dumps(payload, ensure_ascii=False).encode("utf-8")
    raw = _request(
        url,
        data=body,
        method="POST",
        headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {key}",
        },
    )
    try:
        data = json.loads(raw.decode("utf-8"))
    except json.JSONDecodeError as exc:
        raise FarmError(f"non-JSON from {url}: {exc}") from exc
    if not isinstance(data, dict):
        raise FarmError(f"empty listing response from {url}")
    return data

File: packages/openclaw-mybot-farm/src/farm-api.mjs
/**
 * Live mybot.farm API helpers (stalls search + GAF pack download + listing create).
 */

const DEFAULT_BASE = "https://mybot.farm";
const DEFAULT_WORKSPACE_ROOT = "~/.openclaw/farm";
export const USER_AGENT = "openclaw-mybot-farm/0.2.0";
export const DEFAULT_BASE_URL = DEFAULT_BASE;

// Exact category *labels* from web/src/lib/site.ts `categories[].label`.
export const CATEGORY_LABELS = Object.freeze([
  "Lifestyle",
  "Productivity",
  "Coding",
  "Writing",
  "Marketing",
  "Sales",
  "Research",
  "Personal finance",
  "Creative",
  "Music",
  "Education",
  "Ops / admin",
  "Experimental",
]);

const CATEGORY_SET = new Set(CATEGORY_LABELS);
export const LISTING_KINDS = Object.freeze(["agent", "team"]);
const LISTING_KIND_SET = new Set(LISTING_KINDS);
export const AGENT_PACK_FORMAT = "mybot.farm/agent-pack";
export const TEAM_PACK_FORMAT = "mybot.farm/team-pack";
export const MIN_TEAM_MEMBERS = 2;
export const MIN_PAID_PRICE_CENTS = 200;
export const MAX_PRICE_CENTS = 999_900;
export const MAX_PACK_CHARS = 500_000;
export const PRICE_HINT = "Choose Free, or a price between $2.00 and $9,999.00.";

export class FarmError extends Error {
  /**
   * HTTP or pack-shape error from mybot.farm.
   * @param {string} message
   * @param {number | undefined} [status]
   */
  constructor(message, status) {
    super(message);
    this.name = "FarmError";
    this.status = status;
  }
}

export function resolveFarmConfig(pluginConfig) {
  const envUrl = typeof process.env.MYBOT_FARM_URL === "string" ? process.env.MYBOT_FARM_URL.trim() : "";
  const cfg = pluginConfig ?? {};
  const baseUrl = (
    envUrl ||
    (typeof cfg.baseUrl === "string" ? cfg.baseUrl.trim() : "") ||
    DEFAULT_BASE
  ).replace(/\/+$/, "");
  const workspaceRoot =
    (typeof cfg.workspaceRoot === "string" && cfg.workspaceRoot.trim()) || DEFAULT_WORKSPACE_ROOT;
  return { baseUrl, workspaceRoot };
}

/**
 * Seller key: per-call override, else env MYBOT_FARM_API_KEY, else config apiKey.
 * Never log the returned value.
 */
export function resolveApiKey(pluginConfig, override) {
  if (typeof override === "string" && override.trim()) {
    return override.trim();
  }
  const env =
    typeof process.env.MYBOT_FARM_API_KEY === "string" ? process.env.MYBOT_FARM_API_KEY.trim() : "";
  if (env) return env;
  const cfg = pluginConfig ?? {};
  const fromCfg = cfg.apiKey ?? cfg.api_key;
  return typeof fromCfg === "string" ? fromCfg.trim() : "";
}

function shortErrorBody(raw) {
  const text = (raw || "").trim();
  if (!text) return "";
  try {
    const data = JSON.parse(text);
    if (!data || typeof data !== "object" || Array.isArray(data)) {
      return text.slice(0, 200);
    }
    const parts = [];
    for (const key of ["error", "message"]) {
      const value = data[key];
      if (typeof value === "string" && value.trim()) parts.push(value.trim());
    }
    return parts.length ? parts.join(": ").slice(0, 200) : text.slice(0, 200);
  } catch {
    return text.slice(0, 200);
  }
}

async function farmRequest(url, { accept = "application/json", data, headers, method } = {}) {
  const reqHeaders = {
    Accept: accept,
    "User-Agent": USER_AGENT,
    ...(headers ?? {}),
  };
  let res;
  try {
    res = await fetch(url, {
      method: method ?? (data != null ? "POST" : "GET"),
      headers: reqHeaders,
      body: data,
    });
  } catch (err) {
    const reason = err instanceof Error ? err.message : String(err);
    throw new FarmError(`mybot.farm unreachable for ${url}: ${reason}`);
  }
  if (!res.ok) {
    const body = await res.text().catch(() => "");
    const short = shortErrorBody(body);
    throw new FarmError(
      `mybot.farm ${res.status} for ${url}${short ? `: ${short}` : ""}`,
      res.status,
    );
  }
  return Buffer.from(await res.arrayBuffer());
}

async function farmFetch(baseUrl, path) {
  const url = `${baseUrl}${path.startsWith("/") ? path : `/${path}`}`;
  const raw = await farmRequest(url);
  try {
    return JSON.parse(raw.toString("utf8"));
  } catch (err) {
    throw new FarmError(`non-JSON from ${url}: ${err instanceof Error ? err.message : String(err)}`);
  }
}

export async function searchStalls(baseUrl, query, limit) {
  const q = encodeURIComponent(query);
  const data = await farmFetch(baseUrl, `/api/stalls?q=${q}`);
  let stalls = Array.isArray(data.stalls) ? data.stalls : [];
  if (typeof limit === "number" && Number.isFinite(limit) && limit > 0) {
    stalls = stalls.slice(0, Math.floor(limit));
  }
  return {
    stalls,
    count: typeof data.count === "number" ? data.count : stalls.length,
    query: data.query ?? query,
  };
}

export async function getPack(baseUrl, slug) {
  const clean = slug.trim().replace(/^\/+|\/+$/g, "");
  if (!clean) throw new FarmError("slug required");
  const pack = await farmFetch(baseUrl, `/api/packs/${encodeURIComponent(clean)}`);
  if (!pack || typeof pack !== "object") throw new FarmError(`empty pack for ${clean}`);
  if (!pack.slug) pack.slug = clean;
  return pack;
}

export async function getInstallPrompt(baseUrl, slug) {
  const clean = slug.trim();
  return farmFetch(baseUrl, `/api/install-prompt/${encodeURIComponent(clean)}`);
}

export function stallSummary(stall) {
  return {
    slug: stall.slug,
    stallId: stall.stallId ?? stall.listingId ?? "",
    packVersion: stall.packVersion ?? null,
    name: stall.name ?? stall.slug,
    title: stall.title ?? "",
    pageUrl: stall.pageUrl ?? `https://mybot.farm/agents/${stall.slug}`,
    packUrl: stall.packUrl ?? "",
    hermesHref: stall.hermesHref ?? stall.hermesUrl ?? "",
    description: stall.description ?? "",
    category: stall.category ?? "",
  };
}

export function packSummary(pack) {
  const skills = Array.isArray(pack.skills) ? pack.skills : [];
  const manifest = pack.manifest ?? {};
  return {
    slug: pack.slug,
    format: pack.format,
    version: pack.version,
    packVersion: pack.packVersion,
    profile: pack.profile ?? {},
    skillNames: skills.map((s) => s.name).filter(Boolean),
    skillCount: skills.length,
    memoryCount: Array.isArray(pack.memory) ? pack.memory.length : 0,
    attribution: manifest.attribution ?? "",
    sourceNote: manifest.sourceNote ?? "",
    sourceRepo: manifest.sourceRepo ?? "",
    license: manifest.license ?? "",
    homepage: manifest.homepage ?? `https://mybot.farm/agents/${pack.slug}`,
  };
}

/** Match web/src/lib/listings.ts parsePriceCents (plus string ints for CLI). */
export function parsePriceCents(value) {
  if (typeof value === "boolean") return null;
  if (typeof value === "string") {
    const text = value.trim();
    if (!text) return null;
    const parsed = text.includes(".") ? Number(text) : Number.parseInt(text, 10);
    if (!Number.isFinite(parsed)) return null;
    value = parsed;
  }
  if (typeof value !== "number" || !Number.isFinite(value)) return null;
  const cents = Math.round(value);
  if (cents === 0) return 0;
  if (cents < MIN_PAID_PRICE_CENTS || cents > MAX_PRICE_CENTS) return null;
  return cents;
}

export function parseListingKind(value) {
  if (typeof value === "string" && LISTING_KIND_SET.has(value.trim())) {
    return value.trim();
  }
  return null;
}

export function parsePackObject(value) {
  if (typeof value === "string") {
    try {
      value = JSON.parse(value);
    } catch {
      throw new FarmError("Pack JSON must be an object.");
    }
  }
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new FarmError("Pack JSON must be an object.");
  }
  const encoded = JSON.stringify(value);
  if (encoded.length > MAX_PACK_CHARS) {
    throw new FarmError("Pack JSON is too large (max 500 KB).");
  }
  return value;
}

function memberPackError(index, pack) {
  if (typeof pack === "string") {
    return pack.trim()
      ? null
      : `members[${index}].pack must be a catalog path, slug, tarball URL, or nested agent-pack object`;
  }
  if (!pack || typeof pack !== "object" || Array.isArray(pack)) {
    return `members[${index}].pack must be a catalog path, slug, tarball URL, or nested agent-pack object`;
  }
  if (pack.format === TEAM_PACK_FORMAT) {
    return `members[${index}].pack nested object cannot be a team-pack`;
  }
  return null;
}

/** Match web/src/lib/gaf-pack.ts validateListingPack for farm_post dry-run. */
export function validateListingPack(kind, pack) {
  const fmt = typeof pack?.format === "string" ? pack.format.trim() : "";
  if (kind === "team") {
    if (fmt !== TEAM_PACK_FORMAT) {
      throw new FarmError(`kind "team" requires pack.format "${TEAM_PACK_FORMAT}"`);
    }
    const members = pack.members;
    if (!Array.isArray(members) || members.length < MIN_TEAM_MEMBERS) {
      throw new FarmError(
        `kind "team" requires members[] with at least ${MIN_TEAM_MEMBERS} agents`,
      );
    }
    for (let i = 0; i < members.length; i += 1) {
      const member = members[i];
      if (!member || typeof member !== "object" || Array.isArray(member)) {
        throw new FarmError(`members[${i}] must be an object with role, summary, and pack`);
      }
      const role = typeof member.role === "string" ? member.role.trim() : "";
      const summary = typeof member.summary === "string" ? member.summary.trim() : "";
      if (!role) throw new FarmError(`members[${i}].role is required`);
      if (!summary) throw new FarmError(`members[${i}].summary is required`);
      const packError = memberPackError(i, member.pack);
      if (packError) throw new FarmError(packError);
    }
    if (pack.shared !== undefined) {
      if (!pack.shared || typeof pack.shared !== "object" || Array.isArray(pack.shared)) {
        throw new FarmError("shared must be an object.");
      }
      if (
        pack.shared.gettingStarted !== undefined &&
        typeof pack.shared.gettingStarted !== "string"
      ) {
        throw new FarmError("shared.gettingStarted must be a string (Hermes install steps).");
      }
    }
    return;
  }
  if (fmt === TEAM_PACK_FORMAT) {
    throw new FarmError(`kind "agent" cannot use pack.format "${TEAM_PACK_FORMAT}"`);
  }
  if (Array.isArray(pack.members) && pack.members.length > 0) {
    throw new FarmError('kind "agent" listings cannot include members[] — use kind "team"');
  }
}

export function buildListingPayload({
  kind,
  name,
  title,
  description,
  category,
  priceCents,
  pack,
  slug,
  packVersion,
}) {
  const parsedKind = parseListingKind(kind);
  if (!parsedKind) {
    throw new FarmError('kind must be "agent" or "team"');
  }

  const parsedName = typeof name === "string" ? name.trim() : "";
  const parsedTitle = typeof title === "string" ? title.trim() : "";
  const parsedDescription = typeof description === "string" ? description.trim() : "";
  if (!parsedName || !parsedTitle || !parsedDescription) {
    throw new FarmError("name, title, and description are required");
  }

  const parsedCategory = typeof category === "string" ? category.trim() : "";
  if (!CATEGORY_SET.has(parsedCategory)) {
    const labels = [...CATEGORY_LABELS].sort().join(", ");
    throw new FarmError(`category must be an exact farm label (${labels})`);
  }

  const parsedPrice = parsePriceCents(priceCents);
  if (parsedPrice === null) {
    throw new FarmError(PRICE_HINT);
  }

  const parsedPack = parsePackObject(pack);
  validateListingPack(parsedKind, parsedPack);
  const payload = {
    kind: parsedKind,
    name: parsedName,
    title: parsedTitle,
    description: parsedDescription,
    category: parsedCategory,
    priceCents: parsedPrice,
    pack: parsedPack,
  };
  if (typeof slug === "string" && slug.trim()) {
    payload.slug = slug.trim().toLowerCase();
  }
  if (packVersion != null && packVersion !== "") {
    payload.packVersion = packVersion;
  }
  return payload;
}

export function listingPayloadSummary(payload) {
  const pack = payload.pack && typeof payload.pack === "object" ? payload.pack : {};
  const skills = Array.isArray(pack.skills) ? pack.skills : [];
  const members = Array.isArray(pack.members) ? pack.members : [];
  const encoded = JSON.stringify(pack);
  const memberRoles = members
    .filter((m) => m && typeof m === "object" && typeof m.role === "string" && m.role.trim())
    .map((m) => m.role.trim());
  return {
    kind: payload.kind,
    name: payload.name,
    title: payload.title,
    category: payload.category,
    priceCents: payload.priceCents,
    slug: payload.slug,
    packVersion: payload.packVersion,
    pack: {
      format: pack.format,
      version: pack.version,
      packVersion: pack.packVersion,
      runtime: pack.runtime || [],
      skillCount: skills.length,
      memberCount: members.length,
      memberRoles,
      encodedChars: encoded.length,
    },
  };
}

export function listingPageUrl(baseUrl, pagePath) {
  const path = pagePath.startsWith("/") ? pagePath : `/${pagePath}`;
  return `${baseUrl.replace(/\/+$/, "")}${path}`;
}

export async function createListing(baseUrl, payload, apiKey) {
  const key = typeof apiKey === "string" ? apiKey.trim() : "";
  if (!key) {
    throw new FarmError(
      "seller API key required (env MYBOT_FARM_API_KEY or plugin config apiKey)",
    );
  }
  const url = `${baseUrl.replace(/\/+$/, "")}/api/listings`;
  const body = JSON.stringify(payload);
  const raw = await farmRequest(url, {
    data: body,
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${key}`,
    },
  });
  let data;
  try {
    data = JSON.parse(raw.toString("utf8"));
  } catch (err) {
    throw new FarmError(`non-JSON from ${url}: ${err instanceof Error ? err.message : String(err)}`);
  }
  if (!data || typeof data !== "object" || Array.isArray(data)) {
    throw new FarmError(`empty listing response from ${url}`);
  }
  return data;
}
